src/dhtproto/node/neo/request/GetChannels.d

```diff
@@ -1 +1,89 @@
+/*******************************************************************************
+
+    v0 GetChannels request protocol.
+
+    Copyright:
+        Copyright (c) 2017 sociomantic labs GmbH. All rights reserved.
+
+    License:
+        Boost Software License Version 1.0. See LICENSE.txt for details.
+
+*******************************************************************************/
+
 module dhtproto.node.neo.request.GetChannels;
+
+/// ditto
+public abstract scope class GetChannelsProtocol_v0
+{
+    import swarm.neo.node.RequestOnConn;
+    import swarm.neo.connection.RequestOnConnBase;
+    import dhtproto.common.GetChannels;
+    import dhtproto.node.neo.request.core.Mixins;
+    import ocean.core.Array : copy;
+    import ocean.transition;
+
+    /// Mixin the constructor and resources member.
+    mixin RequestCore!();
+
+    /// Connection to the client.
+    private RequestOnConn connection;
+
+    /***************************************************************************
+
+        Request handler. Responds to the client with a status code and sends the
+        list of channels.
+
+        Params:
+            connection = connection to client
+            msg_payload = initial message read from client to begin the request
+                (the request code and version are assumed to be extracted)
+
+    ***************************************************************************/
+
+    final public void handle ( RequestOnConn connection,
+        Const!(void)[] msg_payload )
+    {
+        this.connection = connection;
+
+        // Send status code
+        this.connection.event_dispatcher.send(
+            ( RequestOnConnBase.EventDispatcher.Payload payload )
+            {
+                payload.addConstant(RequestStatusCode.Started);
+            }
+        );
+
+        auto channel_buf = this.resources.getVoidBuffer();
+
+        foreach ( channel; this )
+        {
+            // Copy the channel name in case it changes during sending.
+            (*channel_buf).copy(channel);
+            enableStomping(*channel_buf);
+
+            this.connection.event_dispatcher.send(
+                ( RequestOnConnBase.EventDispatcher.Payload payload )
+                {
+                    payload.addConstant(MessageType.ChannelName);
+                    payload.addArray(*channel_buf);
+                }
+            );
+        }
+
+        // Send finished code
+        this.connection.event_dispatcher.send(
+            ( RequestOnConnBase.EventDispatcher.Payload payload )
+            {
+                payload.addConstant(MessageType.Finished);
+            }
+        );
+    }
+
+    /***************************************************************************
+
+        opApply iteration over the names of the channels in storage.
+
+    ***************************************************************************/
+
+    protected abstract int opApply ( int delegate ( ref cstring ) dg );
```

Comment:
Nitpick: this is a bit odd. I find `opApply` to be a member of the struct we're iterating over, but here it's the iteration over some other entity. I can't think of any better (which is also as easy to implement - obviously, you could declare a base abstract class here, and make an abstract method returning an instance of it, but that's just too much work for very little benefit) solution.

Reply:
I know what you mean. But yeah, the reason I did it like this is for convenience / simplicity.

Reply:
I know what you mean.